### PR TITLE
Fix for #Pydev 1040: Undefined Variable when using below defined classes with __future__ imported

### PR DESCRIPTION
**Pull request by oluiscabral**

### CODE REVIEW

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java [hunk 1]
```diff
@@ -399,7 +406,9 @@ public Object visitFunctionDef(FunctionDef node) throws Exception {
         if (args.annotation != null) {
             for (exprType expr : args.annotation) {
                 if (expr != null) {
+                    startScope(Scope.SCOPE_TYPE_ANNOTATION, node);
```
Comment (fabioz) [inline plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java:409]: The start scope node should be `expr`.

**Review by fabioz — CHANGES_REQUESTED** [review]

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java (hunk 1 above)
Comment (oluiscabral) [reply plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java]: Yup, I mistyped it for some reason...

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java [hunk 2]
```diff
@@ -326,7 +332,8 @@ protected void addToNamesToIgnore(SimpleNode node, boolean finishClassScope, boo
                 //found match in names to ignore...
 
                 if (finishClassScope && foundScopeType == Scope.SCOPE_TYPE_CLASS
-                        && scope.getCurrScopeId() < single.scopeFound.getScopeId()) {
+                        && scope.getCurrScopeId() < single.scopeFound.getScopeId()
+                        || (!futureAnnotationsImported && foundScopeType == Scope.SCOPE_TYPE_ANNOTATION)) {
```
Comment (fabioz) [inline plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java:336]: This if clause seems to be missing some parenthesis...

I think it should be something as the code below (please run all the tests afterwards):

```
(
   finishClassScope && scope.getCurrScopeId() < single.scopeFound.getScopeId() && (
        foundScopeType == Scope.SCOPE_TYPE_CLASS ||
        (!futureAnnotationsImported && foundScopeType == Scope.SCOPE_TYPE_ANNOTATION)
    )
)
```

**Review by fabioz — CHANGES_REQUESTED** [review]

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java (hunk 2 above)
Comment (oluiscabral) [reply plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/scopeanalysis/AbstractScopeAnalyzerVisitor.java]: I ran all tests and this change seems to work fine. Already updated this handle in the last commit ;)

**Comment by vnmabus** [comment]
Is this solution supposed to work also with annotations for the return type? Because it is not working for me.

**Comment by fabioz** [comment]
@vnmabus if it's not working for some specific use case, please create a ticket at: https://www.brainwy.com/tracker/PyDev/ describing what's not working for you.

**Pull request merged**